Add full cvmfsexec support in GWMS

namrathaurs opened the pull request:
This PR supersedes #372 and contains the changes from the same. 

During the implementation and testing of full cvmfsexec support (coming through this PR), a local branch by the name `branch_v3_11_revised_latest` was in use. Since there were a couple of development releases in the meantime, this PR had to be rebased with `branch_v3_11` which resulted in feature-related changes to go out of sync. To avoid further confusion, a new branch was created based on `branch_v3_11_revised_latest` and was named `v311_revised_latest_dupe`, which is the source for this PR.

Dependencies:
- [x] #683

namrathaurs committed:
final changes for full cvmfsexec support

namrathaurs committed:
final changes for full cvmfsexec support

namrathaurs committed:
adding unit tests for on-demand CVMFS provisioning

namrathaurs committed:
adding unit tests for on-demand CVMFS provisioning

shreyb commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -75,7 +75,10 @@ build_cvmfsexec_distros() {
+	# download the cvmfsexec repository contents as a compressed tarball and extract its contents
+	# before extracting contents, make sure the `latest` directory for cvmfsexec exists
+	ensure_directory_exists "$cvmfsexec_latest"
+	curl -sL https://github.com/cvmfs/cvmfsexec/archive/master.tar.gz | tar -xz -C "$cvmfsexec_latest" --strip-components=1
```
Similar to how line 78 had a reference to `$CVMFSEXEC_REPO`, it might make this more readable to set a global variable to this URL, and then reference the variable here.

namrathaurs replied:
Point taken, thanks for the suggestion @shreyb!

shreyb reviewed (commented):
This mostly looks good to me.  I had just one comment.

namrathaurs commented:
Hi Shreyas,

Somehow I could not see this comment of yours on Github, so replying here:

In Bash, the ,, is used to convert the value to lowercase. This helps resolve all unintentional specification of values of `glidein_cvmfs_require` (e.g. Required, REQUIRED, etc.) to required.
Possible values of `GLIDEIN_CVMFS_REQUIRE` (which is a factory config knob) include REQUIRED, PREFERRED, and NEVER.


Hope that helps,
Namratha

________________________________
From: Shreyas Bhat ***@***.***>
Sent: Wednesday, March 11, 2026 9:19 AM
To: glideinWMS/glideinwms ***@***.***>
Cc: Urs, NAMRATHA ***@***.***>; Author ***@***.***>
Subject: [EXT] Re: [glideinWMS/glideinwms] Add full cvmfsexec support in GWMS (PR #631)


@shreyb commented on this pull request.

________________________________

In creation/web_base/glidein_startup_ff.sh<https://github.com/glideinWMS/glideinwms/pull/631#discussion_r2918682476>:

>      mount_cvmfs_repos $gwms_cvmfsexec_mode $cvmfs_config_repo $cvmfs_add_repos
+    if [[ $? -ne 0 ]]; then
+        glidein_cvmfs_require=$(gconfig_get GLIDEIN_CVMFS_REQUIRE "$glidein_config")
+        glidein_cvmfs_require=${glidein_cvmfs_require,,}


What does the ,, do with the parameter?

—
Reply to this email directly, view it on GitHub<https://github.com/glideinWMS/glideinwms/pull/631#pullrequestreview-3929953500>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADY4LAS3ILDONQFPD5F6LOL4QFYW7AVCNFSM6AAAAACP3WHPGOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTSMRZHE2TGNJQGA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

namrathaurs committed:
addressed review comments and minor improvments

- added capability for the script to support only listing of allowed
  platforms based on the support machine types in cvmfsexec.
- added two new functions to dynamically obtain a list of supported
  machine types based on what's supported by cvmfsexec utility.
- updated existing code to use the new functions and indentation inconsistencies for a few lines.
- defined the cvmfsexec archive url as a literal before using it.
- updated the regular expression to include the recently added machine
  types in cvmfsexec and updated the erro handling message to dynamically populate the default list of machine types.
- added documentation to the functions in the script.

namrathaurs committed:
added documentation for functions

mambelli commented on creation/web_base/cvmfsexec_platform_select.sh:
```diff
@@ -20,13 +20,15 @@ add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" |
 
 # this script should run only when on-demand CVMFS is requested by the user/job,
-# so use the GLIDEIN_USE_CVMFSEXEC from the config file as a flag to determine
+# so use the GLIDEIN_USE_CVMFS from the config file as a flag to determine
```
The documentation should be updated.
https://glideinwms.fnal.gov/doc.prd/factory/custom_vars.html#cvmfs_vars should include all CVMFS variables with current names.
The table should be updated: 
- the "cvmfsexec mode used" column should say "No" if CVMFS is not used/mounted (not N/A) and "Error" is the Glidein is failing (e.g. required and unable to provide it.
- Maybe 0/REQUIRED should still mount CVMFS
- What happens with 1/PREFERRED|OPTIONAL and no FUSE and no unpriv userns? That line seems missing

Maybe separate the table in 2: one to define the level of requirement for CVMFS, one for the actions depending on the value of that level

namrathaurs replied:
The documentation link in the comment above (https://glideinwms.fnal.gov/doc.prd/factory/custom_vars.html#cvmfs_vars) is the one for production version. With respect to production version of on-demand CVMFS provisioning (via mode 1), the names of CVMFS variables has NOT CHANGED.

The documentation for development series (https://glideinwms.fnal.gov/doc.dev/factory/custom_vars.html#cvmfs_vars) is being updated via this PR to update the newer names for CVMFS variables. See `#cvmfs_vars` section in `doc/factory/custom_vars.html` file of this PR. If you'd like to preview the updated documentation page (from this PR), you can access it [here](https://htmlpreview.github.io/?https://github.com/namrathaurs/glideinwms/blob/v311_revised_latest_dupe/doc/factory/custom_vars.html#cvmfs_vars).

Some notes:
* Two of the three CVMFS variables have newer names:
    1. `GLIDEIN_USE_CVMFSEXEC` is changing to `GLIDEIN_USE_CVMFS`.
    2. `GLIDEIN_CVMFS` is changing to `GLIDEIN_CVMFS_REQUIRE`.
    3. `CVMFS_SRC` remains unchanged.
* `GLIDEIN_CVMFS_REQUIRE` can only take one of three values: `REQUIRED`, `PREFERRED` or `NEVER` (as opposed to four values which includes the three plus `OPTIONAL` values). Decided to remove `OPTIONAL` from the new list of possible values since both `PREFERRED` and `OPTIONAL` might end up having the same behavior and found the following in current (production) documentation as well:
    
    > REQUIRED - if unable to mount, report an error and abort mounting);
    > **PREFERRED - if unable to mount, notify and continue normally;
    > OPTIONAL - continue if unable to mount (similar to PREFERRED);**
    > NEVER - only exit if mounting fails

<br/>

> Maybe separate the table in 2: one to define the level of requirement for CVMFS, one for the actions depending on the value of that level

I like the suggestion to separate the table into two separate tables and updated to reflect the same. There seems to be redundant information (in the new table and the table containing on-demand CVMFS knobs). I'm retaining the redundancy so that you can evaluate if your suggestion for two tables is still necessary since I already had the information put under the CVMFS knobs table.

mambelli commented on creation/web_base/cvmfsexec_platform_select.sh:
```diff
@@ -20,13 +20,15 @@ add_config_line_source=$(grep -m1 '^ADD_CONFIG_LINE_SOURCE ' "$glidein_config" |
+#use_cvmfs=${use_cvmfs,,}
 
-if [[ "$use_cvmfsexec" -ne 1 ]]; then
+cvmfs_require=$(gconfig_get GLIDEIN_CVMFS_REQUIRE "$glidein_config")
```
GLIDEIN_CVMFS_REQUIRE is not in the documentation. Is it the new name of GLIDEIN_CVMFS?

namrathaurs replied:
Yes, `GLIDEIN_CVMFS_REQUIRE` is the new name of `GLIDEIN_CVMFS`. The new name will not be reflected in the documentation for production version because the production version uses old names. 

The new name has been updated in the documentation for development version, which is coming in as a change through this PR, and as can be seen [here](https://htmlpreview.github.io/?https://github.com/namrathaurs/glideinwms/blob/v311_revised_latest_dupe/doc/factory/custom_vars.html#cvmfs_vars).

mambelli commented on creation/web_base/glidein_startup_ff.sh:
```diff
@@ -2142,8 +2142,20 @@ if [[ -n "$gwms_cvmfs_reexec" && "$gwms_cvmfs_reexec" == "yes" ]]; then
+        if [[ "${glidein_cvmfs_require}" == "required" ]]; then
+            # if mount CVMFS is not successful, report an error and exit with failure exit code
+            logerror "Unable to mount CVMFS on worker node; aborting glidein setup (CVMFS ${glidein_cvmfs_require})"
+            exit 1
```
Should this check and failure be in one of the custom scripts?
Are you sure you can exit without invoking any cleanup function?

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -6,13 +6,58 @@
+        rm -rf "$temp_loc"
+        return 0
+    else
+        error_handler "Something went wrong while cleaning up!"
```
Usually functions are defined before using them (no shell error though).
Then you use error_handler that prints usage but I don't think this is an use error. And you have a return after the exit in the function

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -6,13 +6,58 @@
+## \brief Get all the machine types supported by cvmfsexec utility.
+## \param No parameters.
+## \returnval 0 if the supported machine types were obtained and cleanup of the temporary directory was successful, 1 if the supported machine types were obtained but cleanup of the temporary directory failed.
+# Check with Marco: is the return value necessary? since I'm not using the return value anywhere to determine the next course of action
```
If functions do a task w/o printing error messages or exiting, then the return value is used to tell the caller if there was an error. You need it if it's checked by the caller, you can have it anyway for future uses or clarity in the function. In this case you do a mix of things depending on the path in the function. There is an exit, echo a return and a call to error_handler that does exit and return.

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -6,13 +6,58 @@
+## \returnval No return value.
+set_default_machine_types() {
+    # first, get the machine types supported by the cvmfsexec utility
+    machine_types=$(get_supported_machine_types)
+    # then, process the output to have a comma-separated list of machine types
+    # converting the multi-line output to single line string
+    machine_types=${machine_types//$'\n'/ }
+    # trimming leading/trailing spaces in the single line string
+    machine_types=$(echo "$machine_types" | awk '$1=$1')
+    # replacing spaces with commas
+    machine_types=${machine_types// /,}
+    echo "$machine_types"
```
Can multiple machine types also be printed on the same line space separated by `get_supported_machine_types`?
Is `get_supported_machine_types` adding extra spaces?
If not to both, then something like this may be more efficient:
```
machine_types=$(get_supported_machine_types)
machine_types=${machine_types//$'\n'/,}
echo "${machine_types%,}"
```
The last one, `${%}` is likely not needed because $() is trimming the las newline, but just in case

namrathaurs replied:
> Can multiple machine types also be printed on the same line space separated by get_supported_machine_types?

`get_supported_machine_types` is a custom created function that is solely used to get all the possible machine types supported by cvmfsexec utility. By default (which is native cvmfsexec behavior), machine types are not printed on the same line, and hence causes the list of machine types returned to be a multi-line output string.
```
# output returned by native cvmfsexec
...
...
  rhel7-x86_64
  rhel8-aarch64 rhel8-ppc64le rhel8-x86_64
  rhel9-aarch64 rhel9-ppc64le rhel9-x86_64
  rhel10-x86_64
  suse15-aarch64 suse15-x86_64
...
...
```
<br/>

> Is get_supported_machine_types adding extra spaces?

No, the `get_supported_machine_types()` is not adding extra spaces, but the multi-line output returned by native cvmfsexec contains extra spaces (i.e. leading/trailing spaces). So, designed `set_default_machine_types()` to explicitly handle the multi-line output by removing leading/trailing spaces and separating the values using commas.
<br/>

>If not to both, then something like this may be more efficient:
>```
> machine_types=$(get_supported_machine_types)
> machine_types=${machine_types//$'\n'/,}
> echo "${machine_types%,}"
> ```
Tried the more efficient approach as you suggested but it still contains extra spaces and is still not a purely comma-separated list of values:
```
MYDEBUG (line 1): machine_types =   rhel7-x86_64
  rhel8-aarch64 rhel8-ppc64le rhel8-x86_64
  rhel9-aarch64 rhel9-ppc64le rhel9-x86_64
  rhel10-x86_64
  suse15-aarch64 suse15-x86_64
MYDEBUG (line2) : machine_types =   rhel7-x86_64,  rhel8-aarch64 rhel8-ppc64le rhel8-x86_64,  rhel9-aarch64 rhel9-ppc64le rhel9-x86_64,  rhel10-x86_64,  suse15-aarch64 suse15-x86_64
# output for line 3
  rhel7-x86_64,  rhel8-aarch64 rhel8-ppc64le rhel8-x86_64,  rhel9-aarch64 rhel9-ppc64le rhel9-x86_64,  rhel10-x86_64,  suse15-aarch64 suse15-x86_64
```

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -6,13 +6,58 @@
+CVMFSEXEC_ARCHIVE="https://github.com/cvmfs/cvmfsexec/archive/master.tar.gz"
 DEFAULT_WORK_DIR="/var/lib/gwms-factory/work-dir"
-# TODO: periodically add rhel, suse and other derivatives as supported by cvmfsexec
+# TODO: periodically verify DEFAULT_MACHINE_TYPES to ensure rhel, suse and other derivatives as supported by cvmfsexec are included in the list
 # NOTE: Although rhel9-x86_64 is supported, el7 tools might not work with el9 files (as suggested by Dave Dykstra) as of July 03, 2023
```
Maybe move these notes where you define set_default_machine_types

namrathaurs replied:
I think I can remove the TODO note since `DEFAULT_MACHINE_TYPES` will always be supporting the same list of platforms as cvmfsexec utility. This is because `DEFAULT_MACHINE_TYPES` relies on `set_default_machine_types()` which internally relies on `get_supported_machine_types()` and this always returns the most up-to-date list of supported platforms.

I'm not sure if moving the NOTE item on line 55 to `set_default_machine_types()` would make sense. Maybe we could discuss this further.

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -29,10 +74,13 @@ or a comma-separated list of values from the options {osg|egi|default}.
 PLATFORMS_LIST (optional): indicates machine types (platform- and architecture-based)
 for which distributions is to be built. Can be empty, a single value or a
 comma-separated list of values from the options {rhel9-x86_64|rhel8-x86_64|rhel7-x86_64|suse15-x86_64|rhel8-aarch64|rhel8-ppc64le}.
```
Should the default from `$(set_default_machine_types)` be added?
Is DEFAULT_MACHINE_TYPES, the defaults, all available or both? If not the first, then the list above should be printed and something like "Defaults to all available platforms."

namrathaurs replied:
Good catch! Line 76 has been updated to dynamically use the value in `DEFAULT_MACHINE_TYPES` variable to keep the comma-separated list of values up to date. 

`DEFAULT_MACHINE_TYPES` always contains all available platforms and is used as the default value when `machine_types` parameter is empty/not specified (this parameter comes from the Factory config `cvmfsexec_distros` tag):
```
....
if [[ -z "$2" ]]; then
	machine_types=$DEFAULT_MACHINE_TYPES
elif [[ "$2" =~ $re_mtypes ]]; then
	machine_types=$2
else
	# handle if there were typos in the arguments passed
	error_handler "Invalid platform provided. Must be empty or comma-separated list with (${DEFAULT_MACHINE_TYPES//,/|}), not '$2'"
fi
....
```

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -213,7 +269,7 @@ elif [[ "$2" =~ $re_mtypes ]]; then
 else
 	# handle if there were typos in the arguments passed
-	error_handler "Invalid platform provided. Must be empty or comma-separated list with (rhel9-x86_64|rhel8-x86_64|rhel7-x86_64|suse15-x86_64|rhel8-aarch64|rhel8-ppc64le), not '$2'"
+	error_handler "Invalid platform provided. Must be empty or comma-separated list with (${DEFAULT_MACHINE_TYPES//,/|}), not '$2'"
```
Should this be just a warning and then fail only if the machine type is not found in the current list? This would allow to run if DEFAULT_MACHINE_TYPES gets obsolete. Is DEFAULT_MACHINE_TYPES, the defaults, all available or both?

namrathaurs replied:
`DEFAULT_MACHINE_TYPES` always contains all available platforms and is used as the default value when `machine_types` parameter is empty/not specified (this parameter comes from the Factory config `cvmfsexec_distros` tag when the `create_cvmfsexec_distros.sh` script is internally invoked during Factory reconfig).

mambelli commented on creation/create_cvmfsexec_distros.sh:
```diff
@@ -200,8 +256,8 @@ fi
 re_sources="^((osg|egi|default),)*(osg|egi|default),?$"
-# TODO: update the regex for other machine types upon checking with `makedist -h` periodically
-re_mtypes="^((rhel(7|8|9)|suse15)(-(x86_64|aarch64|ppc64le),?))+$"
+# TODO: update the regex for other/newly supported machine types upon verifying DEFAULT_MACHINE_TYPES or, alternatively, `makedist -h`
+re_mtypes="^((rhel(7|8|9|10)|suse15)(-(x86_64|aarch64|ppc64le),?))+$"
```
Maybe mention where DEFAULT_MACHINE_TYPES is defined to easily compare/update all lines that have to be kept consistent.

namrathaurs replied:
Moved the TODO note from line 259 to where `DEFAULT_MACHINE_TYPES` variable is defined as suggested.

mambelli commented on doc/factory/custom_vars.html:
```diff
@@ -2892,10 +2891,9 @@ <h2><a name="cvmfs_vars"></a>CVMFS Variables</h2>
-          all the entries that are configured in the factory's configuration
-          file.
+          use with the Glidein to allow CVMFS to be provisioned on-demand. These
+          attributes are configured at the Glidein level so that it applies to
+          all the entries that are defined in the Factory's configuration file.
```
Are these only global or could be also defined in the entry?

mambelli commented on doc/factory/custom_vars.html:
```diff
@@ -2916,18 +2914,17 @@ <h2><a name="cvmfs_vars"></a>CVMFS Variables</h2>
-                <i>cvmfsexec</i> to mount CVMFS on the worker noden. A value of
-                1 indicates <i>cvmfsexec</i> will be used to mount CVMFS on the
-                worker node by the glidein.
+                Not set by default. This attribute controls whether the user/job
```
Does not set equates to 0, 1 or a 3rd option?

namrathaurs replied:
By `not set by default`, I meant to convey that the Factory does not have the `GLIDEIN_USE_CVMFS` knob in the Factory config, which is considered that the Factory is not requesting on-demand CVMFS provisioning. Only if the Factory knob exists in the config along with a set value of 0 or 1, it is considered that the Factory is requesting for on-demand CVMFS provisioning.

mambelli commented on doc/factory/custom_vars.html:
```diff
@@ -2938,47 +2935,96 @@ <h2><a name="cvmfs_vars"></a>CVMFS Variables</h2>
             <td>
               <p>
-                Not set by default. This variable controls the origin (source)
+                Not set by default. This attribute controls the origin (source)
```
Does not set equates to default or something different?

mambelli commented on doc/factory/custom_vars.html:
```diff
@@ -2938,47 +2935,96 @@ <h2><a name="cvmfs_vars"></a>CVMFS Variables</h2>
+                Not set by default. This attribute, in conjunction with
+                GLIDEIN_USE_CVMFS (defined in the Frontend), specifies the
+                site/factory policy to be used when provisioning CVMFS. This
+                also helps in determining the behavior of the Glidein behavior
```
the behavior of the Glidein ~~behavior~~

namrathaurs replied:
fixed

mambelli commented on doc/factory/custom_vars.html:
```diff
@@ -2938,47 +2935,96 @@ <h2><a name="cvmfs_vars"></a>CVMFS Variables</h2>
-                OPTIONAL - continue if unable to mount (similar to
-                PREFERRED);<br />
-                NEVER - only exit if mounting fails<br />
+                Not set by default. This attribute, in conjunction with
```
Does not set equates to one of the possible values? 
Then would be nice to add also a link to the table of behaviots in combination w/ GLIDEIN_USE_CVMFS

mambelli commented on test/bats/creation_web_base_cvmfs_helper_funcs.bats:
```diff
@@ -0,0 +1,367 @@
+    export -f getent
+    export -f cat
+
+    run perform_system_check
+    run log_all_system_info
```
I don't understand where you test the outputs/behaviors.
Is the test just to see if the functions run or fail?

namrathaurs replied:
The latter; I believe I added this test to see if the functions `perform_system_check` and `log_all_system_info` run as expected in a pre-requisite full system check workflow. I do have added other tests in the same test script where I test the outputs/behaviors of the two functions that is in this test.
